src/Bicep.Core/TypeSystem/Providers/ResourceTypeProviderFactory.cs

```diff
@@ -42,35 +44,26 @@ public ResultWithDiagnostic<IResourceTypeProvider> GetResourceTypeProviderFromFi
-            var typesTgzPath = Uri.UnescapeDataString(providerDescriptor.TypesBaseUri?.AbsolutePath ?? throw new UnreachableException("the provider directory doesn't exist"));
-            var typesParentPath = Path.GetDirectoryName(typesTgzPath) ?? throw new UnreachableException("the provider directory doesn't exist");
-
-            // compose the path to the OCI manifest based on the cache root directory and provider version
-            var ociManifestPath = Path.Combine(typesParentPath, "manifest");
-            if (!fileSystem.File.Exists(ociManifestPath))
+            IResourceTypeProvider newResourceTypeLoader;
+            try
             {
-                // always exists since provider restore was successful
-                throw new UnreachableException("the provider manifest path doesn't exist");
+                if (providerDescriptor.Name != AzNamespaceType.BuiltInName)
+                {
+                    // Note (asilverman): the line of code below is meant for 3rd party provider resolution logic which is not yet implemented.
+                    throw new NotImplementedException($"Provider {providerDescriptor.Name} not supported.");
+                }
+                newResourceTypeLoader =
+                new AzResourceTypeProvider(
+                    new AzResourceTypeLoader(
+                        OciTypeLoader.FromDisk(fileSystem, providerDescriptor.TypesBaseUri)),
+                        providerDescriptor.Version);
             }
-
-            // Read the OCI manifest
-            var manifestFileContents = fileSystem.File.ReadAllText(ociManifestPath);
-            OciManifest? ociManifest = JsonConvert.DeserializeObject<OciManifest>(manifestFileContents);
-
-            if (ociManifest is null)
+            catch (Exception ex)
             {
-                return new(x => x.ErrorOccurredReadingFile(ociManifestPath)); ;
+                var invalidArtifactException = ex.As<InvalidArtifactException?>() ?? new InvalidArtifactException(ex.Message, ex, InvalidArtifactExceptionKind.NotSpecified);
+                return new(x => x.ArtifactRestoreFailedWithMessage(providerDescriptor.ArtifactReference, invalidArtifactException.Message));
             }
 
-            using var fileStream = fileSystem.File.OpenRead(Path.Combine(typesParentPath, OciTypeLoader.TypesArtifactFilename));
-
-            IResourceTypeProvider? newResourceTypeLoader = providerDescriptor.Name switch
-            {
-                AzNamespaceType.BuiltInName => new AzResourceTypeProvider(new AzResourceTypeLoader(OciTypeLoader.FromTgz(fileStream)), providerDescriptor.Version),
-                // Note(asilverman): the line of code below is meant for 3rd party provider resolution logic which is not yet implemented.
-                _ => throw new NotImplementedException($"Provider {providerDescriptor.Name} not supported."),
-            };
             return new(cachedResourceTypeLoaders[key] = newResourceTypeLoader);
```

Comment:
I raised this as a separate issue (https://github.com/Azure/bicep/issues/12892), but I think you could structure this code better if you addressed this issue in this PR:

```chsarp
return cachedResourceTypeLoaders.GetOrAdd(key, key => {
    if (providerDescriptor.Name != AzNamespaceType.BuiltInName)
    {
        // Note (asilverman): the line of code below is meant for 3rd party provider resolution logic which is not yet implemented.
        throw new NotImplementedException($"Provider {providerDescriptor.Name} not supported.");
    }

    try
    {
        return new(new AzResourceTypeProvider(
            new AzResourceTypeLoader(
                OciTypeLoader.FromDisk(fileSystem, providerDescriptor.TypesBaseUri)),
                providerDescriptor.Version));
    }
    ...
});
```

Reply:
thanks, I agree with this and have fixed it to the best of my abilities.